tests/integration-tests/tests/iam/test_iam.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import logging
+
+import boto3
+import pytest
+from assertpy import assert_that
+from remote_command_executor import RemoteCommandExecutor
+
+from tests.common.assertions import assert_no_errors_in_logs
+
+
+@pytest.mark.usefixtures("os", "instance")
+def test_iam_roles(
+    region, scheduler, common_pcluster_policies, role_factory, pcluster_config_reader, clusters_factory, cluster_model
+):
+    if scheduler == "awsbatch":
+        instance_policies = common_pcluster_policies["awsbatch_instance_policy"]
+        lambda_policies = common_pcluster_policies["awsbatch_lambda_policy"]
+    else:
+        instance_policies = common_pcluster_policies["traditional_instance_policy"]
+        lambda_policies = common_pcluster_policies["traditional_lambda_policy"]
+    cluster_role_name = role_factory("ec2", [instance_policies])
+    lambda_role_name = role_factory("lambda", [lambda_policies])
+    cluster_config = pcluster_config_reader(
+        config_file=cluster_model + ".ini", ec2_iam_role=cluster_role_name, iam_lambda_role=lambda_role_name
+    )
+    cluster = clusters_factory(cluster_config)
+    main_stack_name = "parallelcluster-" + cluster.name
+    cfn_client = boto3.client("cloudformation", region_name=region)
+    lambda_client = boto3.client("lambda", region_name=region)
+    check_no_role_is_created = scheduler != "awsbatch"
+    _check_lambda_role(cfn_client, lambda_client, main_stack_name, lambda_role_name, check_no_role_is_created)
+    # Lambda function is used to wait_for_update. We test update can be finished
+    cluster.update(force=True)
+
+
+def _check_lambda_role(cfn_client, lambda_client, stack_name, lambda_role_name, check_no_role_is_created):
+    """Test lambda role is attached to all Lambda functions in the stack and its substack."""
+    resources = cfn_client.describe_stack_resources(StackName=stack_name)["StackResources"]
```

Comment:
shall we automatically iterate this research in all substacks and only whitelist the roles we know they can't be overridden?

Reply:
I don't really understand your idea. Here I am doing the recursive call for all substacks, but not whitelisting anything. So you are proposing to have a whitelist?